### DISCUSSION TITLE
Expose validation data to context

### DISCUSSION THREAD

**Discussion by dvv** [trpc/trpc#3682] [Q&A] [answered]
Hi!

I would like to log both input and output zod-validation errors.
I found a couple of places that might help me a lot:

- Add 'Input validation failed' [here](https://github.com/trpc/trpc/blob/main/packages/server/src/core/middleware.ts#L121) to distinguish input validation errors from any other validation errors in the business logic handler. This would be consistent with [output validation message](https://github.com/trpc/trpc/blob/main/packages/server/src/core/middleware.ts#L160)

- In case of input validation error I have `rawInput` as invalid data. In case of output validation error I have no `rawOutput`. If we had something like
```js
    try {
      const data = await parse(result.data);
      return {
        ...result,
        data,
      };
    } catch (cause) {
      ctx.rawOutput = result.data //////// <----------- this would allow to refer to failing data in the logging middleware
      throw new TRPCError({
        message: 'Output validation failed',
        code: 'INTERNAL_SERVER_ERROR',
        cause: getCauseFromUnknown(cause),
      });
    }
``` [here](https://github.com/trpc/trpc/blob/main/packages/server/src/core/middleware.ts#L159) thins would be done.

What do you think of that?

TIA

**Comment by KATT** [comment, marked as answer]
You should be able to do this with a middleware

**Reply by brncsk** [reply]
@KATT could you please point me to some sample code on implementing said middleware? Thanks.

**Reply by kamilogorek** [reply]
Middleware is nothing else than a "pass-through" function, that has access to all the data flowing through it.
In your case, it'd be something like:

```js
const validateAndStoreOutput = t.middleware(async ({ ctx, next, rawInput }) => {
  try {
    return next({
      input: await parse(rawInput),
      ctx,
    });
  } catch (cause) {
    ctx.rawOutput = rawInput;
    throw new TRPCError({
      message: 'Output validation failed',
      code: 'INTERNAL_SERVER_ERROR',
      cause: getCauseFromUnknown(cause),
    });
  }
});

// ...in the router
{
  route: publicProcedure.use(validateAndStoreOutput).mutation(/* */)
}
```

**Reply by brncsk** [reply]
@KATT Thanks – I guess I failed to elaborate on my exact use case. My goal was to expose the `ZodError`s generated during ouput validation but `error.cause` was `undefined` both in the express middleware's `onError` handler and `initTRPC`'s `errorFormatter()`. Turns out I was using node 14, while [`error.cause` is only supported from Node 16.9](https://github.com/tc39/proposal-error-cause#implementations).
